portfolio/src/main/java/com/google/sps/data/CalendarGroup.java

```diff
@@ -0,0 +1,132 @@
+   *     user.
+   * @param startTime: start of possible scheduling blocks, of type Instant. This is the earlier of
+   *     working hour start time and the current timestamp.
+   * @param endTime: end of possible scheduling blocks, of type Instant.
+   */
+  public CalendarGroup(List<CalendarEvent> events, Instant startTime, Instant endTime) {
+    if (events == null) {
+      throw new IllegalArgumentException("Calendar Group needs a collection of events");
+    }
+    if (startTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the start of scheduling hours");
+    }
+    if (endTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the end of scheduling hours");
+    }
+    if (startTime.isAfter(endTime)) {
+      throw new IllegalArgumentException("Start time cannot be after end time");
+    }
+    this.events = events;
+    this.startTime = startTime;
+    this.endTime = endTime;
+    this.allFreeTimeRanges = calculateFreeTimeRanges();
+  }
+
+  /**
+   * Calculates a TimeRange list which represents the periods of time that are empty of events and
+   * lie completely inside the possible scheduling hours. Originally written by tomasalvarez,
+   * modified by hollyyuqizheng.
+   */
+  public List<TimeRange> calculateFreeTimeRanges() {
+    Collections.sort(events, sortByEventStartTimeAscending);
+
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    Instant earliestNonScheduledInstant = startTime;
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's scheduling hours' ending time.
+      if (event.getStartTime().isAfter(earliestNonScheduledInstant)
+          && !event.getStartTime().isAfter(endTime)) {
+        possibleTimes.add(
+            TimeRange.fromStartEnd(earliestNonScheduledInstant, event.getStartTime()));
+      }
+      if (earliestNonScheduledInstant.isBefore(event.getEndTime())) {
+        earliestNonScheduledInstant = event.getEndTime();
+      }
+    }
+    // The end of the work hours is potentially never included so we check.
+    if (endTime.isAfter(earliestNonScheduledInstant)) {
+      possibleTimes.add(TimeRange.fromStartEnd(earliestNonScheduledInstant, endTime));
+    }
+    return possibleTimes;
+  }
+
+  /** Getter for the list of all free time ranges. */
+  public List<TimeRange> getFreeTimeRanges() {
+    return allFreeTimeRanges;
+  }
+
+  /**
+   * Deletes a time range from the list of all free times.
+   *
+   * @return the modifed list of free time ranges.
+   */
+  public List<TimeRange> deleteFreeTimeRange(TimeRange timeRange) {
+    boolean exist = false;
+    for (TimeRange freeTimeRange : allFreeTimeRanges) {
```

Comment:
Collections include a .contains() method. I think this entire method can just be:
``` 
if (allFreeTimeRanges.contains(timeRange)) {    
 allFreeTimeRanges.remove(timeRange);    
} else {   
  throw new IllegalArgumentException("This time range does not represent any free time range.");
return allFreeTimeRanges; 
```
Also it seems like the modification methods should not be returning anything, maybe they should be void.

Reply:
I see about the contains method, I didn't realize the equals method is overriding the default one. 
In terms of the modification method, I was thinking that, since the deleting method returns the new collection, I wanted to keep both methods similar, so both would return the newly modified collection.